pkg/kafka/partition/committer.go

```diff
@@ -63,14 +73,51 @@ func newPartitionCommitter(kafkaCfg kafka.Config, admClient *kadm.Client, partit
 			Help:        "The last consumed offset successfully committed by the partition reader. Set to -1 if not offset has been committed yet.",
 			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
 		}),
+		toCommit: atomic.NewInt64(-1),
 	}
 
 	// Initialise the last committed offset metric to -1 to signal no offset has been committed yet (0 is a valid offset).
 	c.lastCommittedOffset.Set(-1)
 
+	if kafkaCfg.ConsumerGroupOffsetCommitInterval > 0 {
+		c.wg.Add(1)
+		ctx, cancel := context.WithCancel(context.Background())
+		c.cancel = cancel
+		go c.autoCommitLoop(ctx)
+	}
+
 	return c
 }
 
+func (r *partitionCommitter) autoCommitLoop(ctx context.Context) {
+	defer r.wg.Done()
+	commitTicker := time.NewTicker(r.kafkaCfg.ConsumerGroupOffsetCommitInterval)
+	defer commitTicker.Stop()
+
+	previousOffset := r.toCommit.Load()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-commitTicker.C:
+			currOffset := r.toCommit.Load()
+			if currOffset == previousOffset {
+				continue
+			}
+
+			if err := r.Commit(ctx, currOffset); err == nil {
```

Comment:
Edit2: my train of thought here. tl;dr: no action needed.

if we're moving state to the ingester and persisting via the wal, I think we need to commit after each batch is pulled & accepted, not on a timer. Otherwise, restarting will replay the wal and then process the offsets that's in the wal already, resulting in duplicate lines in storage (if they're accepted to diff chunks, otherwise `cut()` will dedupe them).

Edit1: looks like you handle this in `partitionCommitter.Stop()`, which reduces the likelihood a problem here. There's still some gap where duplicates can occur (process dying before it can call the stop() method), but
a) this would be the case between a wal write and subsequent offset commit()
b) running on a timer like this should create fewer offset records in the offsets topic compared to updating them after each write.

All in all, I think this is fine 👍 , considering this problem is going to exist as long as we try to move state from one wal (kafka) to another (ingester) and then update the former. Since we can't make that atomic, this does a good enough job minimizing the blast radius.